src/async-wrap-inl.h

```diff
@@ -21,7 +21,8 @@ inline AsyncWrap::AsyncWrap(Environment* env,
       has_async_queue_(false),
       provider_type_(provider) {
   // Check user controlled flag to see if the init callback should run.
-  if (!env->call_async_init_hook())
+  if (!env->using_asyncwrap() ||
+      (!env->call_async_init_hook() && parent != nullptr))
     return;
```

Comment:
Suggestion: break this into two if statements so it's a little easier to parse for humans.